### ISSUE TITLE
Potential new optimisation for IntegerEqualsNode

### ISSUE THREAD

**Issue by default-jamc** [oracle/graal#6914]
# Overview
Hi,

I noticed there may be another optimisation that can be added to IntegerEqualsNode. Comparing a xor expression to `0` is equivalent to checking that the xor operands are equal.

```java
(x ^ y) == 0 |-> (x == y)
```

This optimisation has also been formally verified in an interactive theorem prover (Isabelle/HOL).

## Sample Program
I've written a program to demonstrate the potential speed-up by this optimisation.

```java
/*
  Demonstrate speed-up of (x ^ y) == 0 -> x == y optimisation
**/
public class IntegerEqualsXorTest {

  public static void main(String[] args) {
    int n = 8;
    long time = 0;
    for (int j = 0; j < n; j++) {
      long start = System.nanoTime();
      long sum = 0;
      for (int i = 0; i < 100000; i++) {
        sum += calculate(i);
      }
      System.out.println(sum);
      long end = System.nanoTime();
      time += (end - start);
      System.out.println("Run: " + j + " | Time taken: " + ((end - start) / 1000000) + "ms");
    }
    System.out.println("Average time: " + ((time / 1000000) / n) + "ms");
  }

  static int calculate(int i) {
    int m = 1;
    int s = 1;
    boolean condition = false;
    for (int j = 0; j < 10000; j++) {
      s = s * 6 + j;

      //i = (j / 3) + (i * 7); // 5-10% speedup
      i = s + 1;               // >10x  speedup 

      //condition = ((s ^ i) == 0); // Unoptimised
      condition = (s == i);         // Optimised

      m = (condition) ? 1 : 2;
    }
    return m;
  }
}
```

The optimised version is 5-10% faster than the unoptimised version most of the time. However some values of `i` (e.g., `i = s + 1`) yielded a 10x speedup with the optimisation.

## Unit Test
I've also written a JUnit test to check whether this optimisation is applied:
```java
public class IntegerEqualsNodeTest extends GraalCompilerTest {

    // x ^ y == 0 |-> x == y
    public static boolean xorEquality1(int x, int y) {
        return ((x ^ y) == 0);
    }

    private void checkXorNodes(String methodName) {
        StructuredGraph graph = parseForCompile(getResolvedJavaMethod(methodName));

        System.out.println("Nodes in graph: ");
        System.out.println("Xor: " + graph.getNodes().filter(XorNode.class).count());
        System.out.println("IntegerEquals: " + graph.getNodes().filter(IntegerEqualsNode.class).count());

        createCanonicalizerPhase().apply(graph, getProviders());

        System.out.println("Nodes in graph: ");
        System.out.println("Xor: " + graph.getNodes().filter(XorNode.class).count());
        System.out.println("IntegerEquals: " + graph.getNodes().filter(IntegerEqualsNode.class).count());

        // If there is no xor node in the graph, the optimisation was applied
        assertTrue(graph.getNodes().filter(XorNode.class).count() == 0);
    }

    @Test
    public void testXor1() {
        checkXorNodes("xorEquality1");
    }
}
```
Running this test produces the following output:

```java
MxJUnitCore
JUnit version 4.12
.Nodes in graph: 
Xor: 1
IntegerEquals: 1
Nodes in graph: 
Xor: 1
IntegerEquals: 1
E
testXor1(org.graalvm.compiler.core.test.IntegerEqualsNodeTest)
java.lang.AssertionError

Time: 0.52
There was 1 failure:
1) org.graalvm.compiler.core.test.IntegerEqualsNodeTest#testXor1
java.lang.AssertionError

FAILURES!!!
Tests run: 1,  Failures: 1
```

**Comment by davleopo** [comment]
thanks @default-jamc  we will have a look, tracked internally as GR-46914

**Comment by rmosaner** [comment]
fixed by #6971
